### PR TITLE
fix(behavior_velocity_intersection_module): behavior_velocity_intersection_module dies due empty conflicting_lanelets

### PR DESCRIPTION
**Pull request by beyzanurkaya**
## Description

https://github.com/autowarefoundation/autoware.universe/pull/3947
Sometimes `conflicting_lanelets` can be empty so can not calculate `adjacent_followings`  and `behavior_velocity_intersection_module` dies.

## Tests performed

Not applicable.

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_velocity_intersection_module/src/util.cpp [hunk 1]
```diff
@@ -468,7 +478,9 @@ IntersectionLanelets getObjectiveLanelets(
         }
         detection_lanelets.push_back(conflicting_lanelet);
         if (!adjacent_followings.empty()) {
-          detection_lanelets.push_back(adjacent_followings.front());
+          for (const auto & adjacent_following : adjacent_followings) {
+            detection_lanelets.push_back(adjacent_following);
+          }
```
Comment (soblin) [inline planning/behavior_velocity_intersection_module/src/util.cpp:483]: @beyzanurkaya Does this snippet need to be inside the for-loop `for (const auto & conflicting_lanelet : conflicting_lanelets)` ? If this snippet is not outside of it, `detection_lanelets` variable will contain duplicate elements in `adjacent_followings` by the number of `conflicting_lanelets().size`.

File: planning/behavior_velocity_intersection_module/src/util.cpp [hunk 2]
```diff
@@ -446,8 +446,18 @@ IntersectionLanelets getObjectiveLanelets(
...
+  if (!conflicting_lanelets.empty()) {
+    for (const auto & conflicting_lanelet : conflicting_lanelets) {
+      for (const auto & following_lanelet : routing_graph_ptr->following(conflicting_lanelet)) {
+        adjacent_followings.push_back(following_lanelet);
+      }
+      for (const auto & following_lanelet : routing_graph_ptr->previous(conflicting_lanelet)) {
+        adjacent_followings.push_back(following_lanelet);
+      }
+    }
+  }
```
Comment (soblin) [inline planning/behavior_velocity_intersection_module/src/util.cpp:460]: You do not need to check `conflicting_lanelets.empty()`

Comment (beyzanurkaya) [reply planning/behavior_velocity_intersection_module/src/util.cpp:458]: @soblin I solved these

**Comment by soblin** [comment]
@beyzanurkaya Thanks! I also suggest this patch:

```
commit 17e343e7d04abc5d09081ca240361b5750f8d651
Author: Mamoru Sobue <mamoru.sobue@tier4.jp>
Date:   Fri Jul 7 19:14:46 2023 +0900

    fix invalid access
    
    Signed-off-by: Mamoru Sobue <mamoru.sobue@tier4.jp>

diff --git a/planning/behavior_velocity_intersection_module/src/scene_intersection.cpp b/planning/behavior_velocity_intersection_module/src/scene_intersection.cpp
index bc6b54c23e..002c4daf4e 100644
--- a/planning/behavior_velocity_intersection_module/src/scene_intersection.cpp
+++ b/planning/behavior_velocity_intersection_module/src/scene_intersection.cpp
@@ -899,6 +899,7 @@ bool IntersectionModule::checkCollision(
     const auto object_direction = util::getObjectPoseWithVelocityDirection(object.kinematics);
     const auto is_in_adjacent_lanelets = util::checkAngleForTargetLanelets(
       object_direction, adjacent_lanelets, planner_param_.common.attention_area_angle_thr,
+      planner_param_.common.consider_wrong_direction_vehicle,
       planner_param_.common.attention_area_margin);
     if (is_in_adjacent_lanelets) {
       continue;
@@ -913,12 +914,14 @@ bool IntersectionModule::checkCollision(
       } else if (util::checkAngleForTargetLanelets(
                    object_direction, attention_area_lanelets,
                    planner_param_.common.attention_area_angle_thr,
+                   planner_param_.common.consider_wrong_direction_vehicle,
                    planner_param_.common.attention_area_margin)) {
         target_objects.objects.push_back(object);
       }
     } else if (util::checkAngleForTargetLanelets(
                  object_direction, attention_area_lanelets,
                  planner_param_.common.attention_area_angle_thr,
+                 planner_param_.common.consider_wrong_direction_vehicle,
                  planner_param_.common.attention_area_margin)) {
       // intersection_area is not available, use detection_area_with_margin as before
       target_objects.objects.push_back(object);
diff --git a/planning/behavior_velocity_intersection_module/src/util.cpp b/planning/behavior_velocity_intersection_module/src/util.cpp
index a3e6689515..17c1a6d3d0 100644
--- a/planning/behavior_velocity_intersection_module/src/util.cpp
+++ b/planning/behavior_velocity_intersection_module/src/util.cpp
@@ -1001,7 +1001,8 @@ Polygon2d generateStuckVehicleDetectAreaPolygon(
 
 bool checkAngleForTargetLanelets(
   const geometry_msgs::msg::Pose & pose, const lanelet::ConstLanelets & target_lanelets,
-  const double detection_area_angle_thr, const double margin)
+  const double detection_area_angle_thr, const bool consider_wrong_direction_vehicle,
+  const double margin)
 {
   for (const auto & ll : target_lanelets) {
     if (!lanelet::utils::isInLanelet(pose, ll, margin)) {
@@ -1012,7 +1013,7 @@ bool checkAngleForTargetLanelets(
     const double angle_diff = tier4_autoware_utils::normalizeRadian(ll_angle - pose_angle);
     if (
       std::fabs(angle_diff) < detection_area_angle_thr / 2 ||
-      std::fabs(angle_diff) > detection_area_angle_thr) {
+      (consider_wrong_direction_vehicle && std::fabs(angle_diff) > detection_area_angle_thr)) {
       return true;
     }
   }
diff --git a/planning/behavior_velocity_intersection_module/src/util.hpp b/planning/behavior_velocity_intersection_module/src/util.hpp
index 89931a5db1..8a539f67f7 100644
--- a/planning/behavior_velocity_intersection_module/src/util.hpp
+++ b/planning/behavior_velocity_intersection_module/src/util.hpp
@@ -144,7 +144,8 @@ Polygon2d generateStuckVehicleDetectAreaPolygon(
 
 bool checkAngleForTargetLanelets(
   const geometry_msgs::msg::Pose & pose, const lanelet::ConstLanelets & target_lanelets,
-  const double detection_area_angle_thr, const double margin = 0.0);
+  const double detection_area_angle_thr, const bool consider_wrong_direction_vehicle,
+  const double margin = 0.0);
 
 void cutPredictPathWithDuration(
   autoware_auto_perception_msgs::msg::PredictedObjects * objects_ptr,

```

Since the feature of considering wrong direction NPCs should only be used when the flag is true, the flag should also be passed `checkAngleForTargetLanelets` function.

**Comment by beyzanurkaya** [comment]
@soblin Thanks for the suggestion! I added the related flag in the `checkAngleForTargetLanelets` function.

**Pull request merged**